### DISCUSSION TITLE
More guidance writing Node formulae

### DISCUSSION THREAD

**Discussion by daxelrod** [Homebrew/discussions#2804] [Tap maintenance and Homebrew development]
Hi! The [tap onboarding experience with `brew tap-new`](https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap) is incredibly easy and well documented! However, after that, I ran into some roadblocks creating a formula for a Node command-line utility following [Node for Formula Authors](https://docs.brew.sh/Node-for-Formula-Authors). I'd like to file a PR to add some details to the docs, but first I want to make sure I solved things in the right way.

When following the example, I ran into the following problems:

1. `node` is not necessarily in `$PATH`(perhaps because I depend on `node@16`?), causing a shebang of `#!/usr/bin/env node` to fail to find an interpreter
2. Using `inreplace` to rewrite the shebang and then `bin.install_symlink Dir["#{libexec}/bin/*"]` resulted in the code not finding its node_modules directory correctly when calling `require()`, for reasons I still don't quite understand (between brew and node, there are 4 levels of symlinks in play). Happy to ressurect a reproducer here if that would be helpful.

[Here's the working formula](https://github.com/daxelrod/homebrew-jowl/blob/5fe59e1c0fda61a9892b995740d6953948ae47e4/Formula/jowl.rb) I ended up with. It uses `write_env_script` to add the Node formula's `opt_bin` to the PATH and then executes `libexec/"bin/jowl"`.

Is the way I handled this recommended practice? If not, what's the right way to do this?

I also attempted to look for prior art in homebrew-core, and there appears to be quite a diversity of approaches:
* [node-sass](https://github.com/Homebrew/homebrew-core/blob/master/Formula/node-sass.rb) uses exactly the format recommended in the docs
* [code-server](https://github.com/Homebrew/homebrew-core/blob/master/Formula/code-server.rb) uses approximately the approach I mentioned above that worked for me
* [yarn](https://github.com/Homebrew/homebrew-core/blob/master/Formula/yarn.rb) does a combination of inreplacing the shebang and writing a wrapper script

Thank you for your guidance here! Once I understand the correct approach, I'll write up a PR to the Node for Formula Authors docs to recommend it explicitly.

**Comment by gromgit** [comment]
Not a maintainer, but there's a reason for the disparate approaches used by the formulae you list:

> `node-sass` uses exactly the format recommended in the docs

This depends on the main `node` formula, which is guaranteed to link its files into the appropriate `HOMEBREW_PREFIX` directories, so no `PATH` adjustments are needed.

> `code-server` uses approximately the approach I mentioned above that worked for me

This, like your own formula, depends on a versioned `node` formula. Versioned formulae are (as far as I know) _never_ linked into `HOMEBREW_PREFIX` on installation, since they _will_ conflict with the main formula that may be installed, either manually or as a dependency of other formulae. This is why you'll see such `PATH`-embedding stuff in many such formulae.

> `yarn` does a combination of inreplacing the shebang and writing a wrapper script

That's a special case, in that (I believe) it's meant to be used with whatever `node` installation is in `$PATH`, so it doesn't have an explicit `node` dependency. The somewhat convoluted `install` block just works around `yarn` peculiarities.

> I'd like to file a PR to add some details to the docs, but first I want to make sure I solved things in the right way.

Since it parallels `code-server`'s approach, I think it should be OK, but the rule of thumb is to put up the PR anyway for discussion.

**Reply by daxelrod** [reply]
Thank you, this brings a bunch of coherency to the situation; in particular that versioned formulas are keg-only but the unversioned `node` is not.

> This depends on the main node formula, which is guaranteed to link its files into the appropriate HOMEBREW_PREFIX directories, so no PATH adjustments are needed.

By unlinking `node`, it is possible to break `sass`:

```
$ brew install node
...
==> Summary
🍺  /home/linuxbrew/.linuxbrew/Cellar/node/17.4.0: 1,983 files, 44.7MB
...
$ brew unlink node
Unlinking /home/linuxbrew/.linuxbrew/Cellar/node/17.4.0... 6 symlinks removed.
$ brew install node-sass
...
==> Pouring node-sass--1.48.0.x86_64_linux.bottle.tar.gz
🍺  /home/linuxbrew/.linuxbrew/Cellar/node-sass/1.48.0: 139 files, 5.4MB
...
$ sass
/usr/bin/env: 'node': No such file or directory
```

So, this leads to a more fundamental question: should formulae _ever_ be written to assume that their dependencies will be in `$PATH`? Are there docs that cover this explicitly?

It looks like common practice for Python is to use [`rewrite_shebang detected_python_shebang`](https://rubydoc.brew.sh/Language/Python/Shebang.html#detected_python_shebang-class_method) and for Perl [`rewrite_shebang detected_perl_shebang`](https://rubydoc.brew.sh/Language/Perl/Shebang.html#detected_perl_shebang-class_method) both of which use the opt_bin of the interpreter formula n the resulting shebang, which should work even if the interpreter isn't linked. That's some prior art at least.

**Reply by SMillerDev** [reply]
> So, this leads to a more fundamental question: should formulae ever be written to assume that their dependencies will be in $PATH?

If a formula depends on something it can depend on that formula being linked.

> It looks like common practice for Python is to use rewrite_shebang detected_python_shebang [..] which use the opt_bin of the interpreter formula n the resulting shebang, which should work even if the interpreter isn't linked. That's some prior art at least.

This is because most software has a shebang that refers to `/usr/bin/python` so we have to rewrite it anyway. Using the opt location is a safer choice in that case.

**Reply by daxelrod** [reply]
Thank you for weighing in!

> If a formula depends on something it can depend on that formula being linked.

With the exception of the dependency formula being a keg-only formula, right?

Ok, that's helpful that formulae aren't expected to work if their dependencies are manually unlinked (is that explicitly documented somewhere, other than maybe checks that `brew doctor` does?)

> This is because most software has a shebang that refers to /usr/bin/python so we have to rewrite it anyway. Using the opt location is a safer choice in that case.

Ohhhh, that makes a ton of sense.

So it sounds like the only change the docs might need is a recommendation for how to handle scripts that depend upon versioned Node so that they can find the interpreter. There are a couple of options here, is there a preferred one?

1. Formulae dependent on versioned Node should use `wite_env_script` to make a wrapper
2. Formulae dependent on versioned Node should use `rewrite_shebang detected_node_shebang` (I'd also create a PR to add a `Language::Node::Shebang#detected_node_shebang`)

Once I know the preferred option. I'll write up the necessary PRs to document it.
